scrape/manager.go

```diff
@@ -223,10 +223,18 @@ func (m *Manager) TargetsActive() map[string][]*Target {
 	m.mtxScrape.Lock()
 	defer m.mtxScrape.Unlock()
 
+	var wg sync.WaitGroup
 	targets := make(map[string][]*Target, len(m.scrapePools))
+	wg.Add(len(m.scrapePools))
 	for tset, sp := range m.scrapePools {
-		targets[tset] = sp.ActiveTargets()
+		// Running in parallel limits the blocking time of scrapePool to scrape
+		// interval when there's an update from SD.
+		go func(tset string, sp *scrapePool) {
+			targets[tset] = sp.ActiveTargets()
```

Comment:
Do we need some locking here?

Reply:
Did you have a chance to look at this?

Reply:
Yes, I will finish it  soon.

Reply:
I might be wrong but isn't this a case of https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables ?

```
go func(tset string, sp *scrapePool) {
 	...
}(tset, &m.scrapePools[tset])
```

https://play.golang.org/p/FTB20cdWLXu

Reply:
The code looks fine to me

Reply:
I think we do need locking. As far as I understand you can't have several goroutines writing in the same map at the same time (even if they modify different keys of the map).

Reply:
The locking has already been added

Reply:
> The code looks fine to me

It looks fine but the goroutines are stacked and can be started after the end of the iteration of the `for loop` (especially if there is nothing else in the for loop body that the goroutine calls).

If you pass a pointer as argument of the goroutine function closure you might end up on calling all the goroutines with the same pointer, the one of the last element of the map you iterate on.

See the following example: https://play.golang.org/p/FTB20cdWLXu

Reply:
We're not passing `&sp` though, we're passing `sp`

Reply:
Indeed, it works in this case: https://play.golang.org/p/XuBiUEu59ci